### DISCUSSION TITLE
OAuth for external services

### DISCUSSION THREAD

**Discussion by kevmcdonk** [microsoft/teams-sdk#171] [Q&A] [answered]
Is there any guidance on how to integrate authentication for other services into the app? For standard Bot Framework Bots, there are options using things TeamsFx and Dialogs but not sure how well that would work with Teams AI where the Bot doesn't get directly called. Is there a way to build Dialogs or an alternative way to gather a token?

**Comment by Stevenic** [comment, marked as answer]
@kevmcdonk if you're used to using the `OAuthPrompt` in your bot, you can actually pass in the settings for the OAuthPrompt in the [authentication](https://github.com/microsoft/teams-ai/blob/stevenic/oauthprompt/js/packages/teams-ai/src/Application.ts#L69) field when configuring your `Application` instance. I'll automatically call the OAuthPrompt on your behalf under the hood and the token will be accessible in the [state.temp.value.authToken field](https://github.com/microsoft/teams-ai/blob/stevenic/oauthprompt/js/packages/teams-ai/src/DefaultTurnStateManager.ts#L43).

This code is deployed as part of the preview.2 but's but the only reason the PR hasn't been merged yet is there's a bug that it doesn't currently replay the original message after successful login. It works you'll just need to send your original message again after you signin. I have some more work to do to fix that.

It actually even support multi-user signin when in group conversations so once I'm done it should be a generally improved signin experience across the board...

**Reply by kevmcdonk** [reply]
Oh that sounds perfect, will give that a try now!

**Reply by kevmcdonk** [reply]
@Stevenic After a little swearing, I've got the OAuth working which is fantastic. Is there a way to access the state field so I can instantiate an object once the auth is done? I have a service and want to mass the token to that rather than have it instantiate in the function each time. I could just check to see if the service has been set up and set if not but feels cleaner to set up at point of authentication if possible.

**Reply by corinagum** [reply]
Hi @kevmcdonk, could you outline your app flow in this scenario to get a tiny bit more context? Presumably, by state I'm guessing you don't mean the bot's `ApplicationState` (Conversation, User, Temp).

------

In case you do mean such, there are options such as

1. Auth token in `temp` state (or user state, as applicable) - saved as a separate value from `authToken` to prevent the call to OAuth
2. Auth token's timer in state

You are able to create any state items you require. For example, you could write
`state.user.value.tokentimer = <timeuntiltokenexpires>`

This will be available to you wherever state is available in the app.
The default available are:
`state.conversation(.value).___`
`state.user(.value).___`
`state.temp(.vallue).___`

The reason why I put value in parenthesis is that one of our future work items is to remove value as a required key to write to state. #49  Once checked in, this will be a breaking change. 

Please let me know if I'm missing the mark on your question. :)

**Reply by kevmcdonk** [reply]
Here's the scenario I've been working through. 
https://github.com/kevmcdonk/TeamsAI-GraphSample/blob/main/index.ts
You can see in "readMail" action (currently on line 154 but may change!) that I set up the graph service and then do the same again with the "listSites" action (currently on line 184 but may change!). It feels wrong to continually be setting up the graph service each time and having the token validated there and that it would make sense to have outside of those and the service re-used. 

THanks for the heads up on the value being removed later.

**Comment by corinagum** [comment]
@kevmcdonk, because using graph service is dependent on helper code within the app, unfortunately there's no way to avoid creating a new GraphService every turn. (`state.temp`) is wiped after every turn, so the authToken won't persist for future calls.
----

There are a few (untested by me) things to potentially try out: 

```ts
app.turn('beforeTurn' (context, state, ...) => {
    //  instantiate `graphService` and save to temporary state; 

    return true;
})

app.action('readMail'...) () => {
    // Access `state.temp.value.graphService` to use the saved state. and call the get mail function
})`

```
This means, however, that `graphService` will be instantiated on every turn, not just `readMail` and `listSites`.

--------

Another idea is [app.chain](https://github.com/microsoft/teams-ai/tree/main/js/samples/04.ai.e.chainedActions.devOpsBot) to perform multiple actions within one turn, though this scenario isn't quite what `app.chain` is designated for.

------

I may have some notes from Steve on using `app.chain` / OAuth, but unfortunately it will take me more time to dig out those notes. In the meantime, please take a look at the above, and @Stevenic perhaps will have the chance to chime in.
